src/OrbitGl/App.cpp

```diff
@@ -2302,42 +2448,131 @@ void OrbitApp::AddSymbols(const ModuleIdentifier& module_id,
                           const orbit_grpc_protos::ModuleSymbols& module_symbols) {
   ORBIT_SCOPE_FUNCTION;
   ModuleData* module_data = GetMutableModuleByModuleIdentifier(module_id);
+  // In case fallback symbols were previously loaded, remove them. Careful to call this before
+  // ModuleData::AddSymbols, as it will clear the fallback symbols from the ModuleData, and
+  // FunctionsDataView contains pointers to them.
+  functions_data_view_->RemoveFunctionsOfModule(module_data->file_path());
   module_data->AddSymbols(module_symbols);
 
   const ProcessData* selected_process = GetTargetProcess();
   if (selected_process != nullptr &&
       selected_process->IsModuleLoadedByProcess(module_data->file_path())) {
     functions_data_view_->AddFunctions(module_data->GetFunctions());
-    ORBIT_LOG("Added loaded function symbols for module \"%s\" to the functions tab",
+    ORBIT_LOG("Added loaded function symbols for module \"%s\" to the Functions tab",
+              module_data->file_path());
+  }
+
+  FireRefreshCallbacks(DataViewType::kModules);
+  UpdateAfterSymbolLoadingThrottled();
+}
+
+void OrbitApp::AddFallbackSymbols(const ModuleIdentifier& module_id,
+                                  const orbit_grpc_protos::ModuleSymbols& fallback_symbols) {
+  ORBIT_SCOPE_FUNCTION;
+  ModuleData* module_data = GetMutableModuleByModuleIdentifier(module_id);
+  module_data->AddFallbackSymbols(fallback_symbols);
+
+  const ProcessData* selected_process = GetTargetProcess();
+  if (selected_process != nullptr &&
+      selected_process->IsModuleLoadedByProcess(module_data->file_path())) {
+    functions_data_view_->AddFunctions(module_data->GetFunctions());
+    ORBIT_LOG("Added fallback symbols for module \"%s\" to the Functions tab",
               module_data->file_path());
   }
 
   FireRefreshCallbacks(DataViewType::kModules);
   UpdateAfterSymbolLoadingThrottled();
 }
 
+Future<ErrorMessageOr<void>> OrbitApp::LoadSymbolsOrFallbackSymbols(
+    const std::filesystem::path& local_file_path, const ModuleIdentifier& module_id) {
+  orbit_base::Future<ErrorMessageOr<void>> load_symbols_future =
+      LoadSymbols(local_file_path, module_id);
+
+  auto load_fallback_symbols =
+      [this, module_id, local_file_path](
+          const ErrorMessageOr<void>& load_debug_result) -> Future<ErrorMessageOr<void>> {
+    if (load_debug_result.has_value()) return {load_debug_result};
+
+    const ModuleData* const module_data = GetModuleByModuleIdentifier(module_id);
+    if (module_data == nullptr) {
+      return {ErrorMessage{absl::StrFormat("Module \"%s\" was not found.", module_id.file_path)}};
+    }
+
+    // Report the error if loading debug symbols fails when the fallback symbols are already loaded.
+    if (module_data->AreAtLeastFallbackSymbolsLoaded()) {
```

Comment:
Will this happen? Should this be an `ORBIT_CHECK`?

Reply:
Ok, I wrote that wrong. The state has to be "Partial". Correcting...